### ISSUE TITLE
Separate resource for Composer extra Python dependencies

### ISSUE THREAD

**Issue by akerone** [hashicorp/terraform-provider-google#12781]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

TL;DR: The installation of extra pypi packages on a Composer environment should be a separate resource.

When creating a Composer environment using Terraform, one of the worst case scenarios is actually one of the most common: Create an environment, wait for 30 to 60 minutes for it to spin up, and watch it fail because there's a conflict in Python dependencies. Rinse and repeat until you find the perfect combination that doesn't break _for this specific image of composer_.

If that happens, the environment is actually created in GCP but tainted in terraform, so the next `apply` has to first destroy the environment and then create it again. It would be much nicer to just try and update the python dependencies on that already created environment.

Additionally, this new resource would be in charge of doing the necessary configuration to install packages from outside pypi, like a specific repository or an artifact registry.

### New or Affected Resource(s)

* google_composer_environment
* google_composer_extra_packages (?)

### Potential Terraform Configuration

```tf
resource "google_composer_extra_packages" "python_dependencies" {
  extra_packages = {
    apache-airflow-providers-sftp = "==4.1.0"

    prettytable                    = "==3.4.1"
    apache-airflow-providers-slack = "==6.0.0"

    # Required for S3 connectivity
    apache-airflow-providers-amazon = "==6.0.0"
    boto3                           = "==1.24.89"
  }

  # URLs taken from google's documentation https://cloud.google.com/composer/docs/how-to/using/installing-python-dependencies
  extra_index_urls = [
    "https://192.0.2.10/",
    "https://us-central1-python.pkg.dev/example-project/example-repository/simple/",
    "https://example.com/"
  ]
}
```

### References

* #10850
  * The new resource could include the configuration to [install from a private repository](https://cloud.google.com/composer/docs/how-to/using/installing-python-dependencies#install-private-repo).

**Comment by rileykarson** [comment]
This could make sense to add, but we're not sure. Adding to `Backlog`.

**Comment by YounesCoder** [comment]
Faced this issue so many times.. it is definitely a good idea

**Comment by akerone** [comment]
By the way, I tried (and failed) to recreate this feature using a local-exec provisioner. It works the first time, but any subsequent applies uninstall them again since they're not in the `software_config` section.

```terraform
resource "local_file" "requirements" {
  filename = "/tmp/requirements.txt"
  content = join("\n", [for k, v in var.extra_pypi_packages : "${k}${v}"])
}

resource "null_resource" "install_pypi_packages" {
  triggers = {
    packages_map = "${md5(jsonencode(var.extra_pypi_packages))}#${md5(google_composer_environment.composer_env.config.0.airflow_uri)}"
  }

  provisioner "local-exec" {
    interpreter = ["bash", "-c"]
    command = <<-EOT
      gcloud composer environments update ${google_composer_environment.composer_env.name} \
      --location ${google_composer_environment.composer_env.region} \
      --update-pypi-packages-from-file ${local_file.requirements.filename}
    EOT
  }
}
```

**Comment by mkumarpcln** [comment]
This is incredibly useful and a must-have IMO. Most (if not all) of our composer updates via terraform fail due to this.

**Comment by CaliWee** [comment]
I believe this issue is solved by: https://github.com/hashicorp/terraform-provider-google/issues/14586

**Comment by akerone** [comment]
Hi again! I'm sorry for the late response.

I don't think this issue is actually solved. The main discussion here is to have the composer environment not be marked as tainted by pip dependency issues, not how to integrate private repositories.

A separate resource for the extra packages is still the best solution in my eyes, as `gcloud` itself creates the environment in those 2 steps even though they packaged it in a single command.

**Comment by CaliWee** [comment]
Apologies, I misread your original question. I agree, the linked issue does not solve your problem

have you considered managing packages at the Task level with [PythonVirtualEnv operators](https://airflow.apache.org/docs/apache-airflow/stable/howto/operator/python.html#pythonvirtualenvoperator)? 

Not a solution, but perhaps a workaround. Alternatively theres also docker + kubernetes pod operators but that is a bit more overhead

**Comment by akerone** [comment]
> have you considered managing packages at the Task level with [PythonVirtualEnv operators](https://airflow.apache.org/docs/apache-airflow/stable/howto/operator/python.html#pythonvirtualenvoperator)?
> 
> Not a solution, but perhaps a workaround. Alternatively theres also docker + kubernetes pod operators but that is a bit more overhead

Those are indeed the two "workarounds" that we're using. Actually, Composer is horrible with dependencies due to the immense amount of packages included with each version to the point where it is almost impossible to install the latest version of anything else. I think any big deployment will have to use PythonVirtualEnv and/or KubernetesPod at some point.

Despite that, there may still be the need for smaller tasks that require extra packages, and not even latest versions of those. That's where I think that the extra PyPi packages play a role as the simplest possible solution. For them to not become a nightmare for platform administrators though, the current environment creation mechanism needs to change.

As @mkumarpcln said, the best current workaround is to first create the environment and then add the packages once the environment is live, which is what I'd do if I was creating it through the Google Cloud Console. In terraform, it becomes a major pain and I think it can be "easily" improved.

I apologize if in my communications I have led to the impression that this is a blocking issue. I'm instead proposing an enhancement that would better reflect in terraform how Google Cloud itself creates composer environments.

Cheers!

**Comment by rileykarson** [comment]
Note: If we address this we'll need to ensure that the relevant field(s) in the complete resource don't detect drift on changes by the separate fine-grained resource for the dependencies.

**Comment by akerone** [comment]
Hi @rileykarson

There are precedents in terraform of this kind of relationship. This should be no different than the relationship between `google_bigquery_dataset` and `google_bigquery_dataset_access`. 

Thanks!
